core/kotlinx-coroutines-test/src/internal/MainTestDispatcher.kt

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2016-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.test.internal
+
+import kotlinx.coroutines.*
+import kotlinx.coroutines.internal.*
+import kotlin.coroutines.*
+
+/**
+ * The testable main dispatcher used by kotlinx-coroutines-test.
+ * It is a [MainCoroutineDispatcher] which delegates all actions to a settable delegate.
+ */
+internal class TestMainDispatcher(private val mainFactory: MainDispatcherFactory) : MainCoroutineDispatcher(), Delay {
+    private var _delegate: CoroutineDispatcher? = null
+    private val delegate: CoroutineDispatcher get() {
+        if (_delegate != null) return _delegate!!
+
+        val newInstance = createDispatcher()
+        if (newInstance != null) {
+            _delegate = newInstance
+            return newInstance
+        }
+
+        // Return missing dispatcher, but do not set _delegate
+        return mainFactory.tryCreateDispatcher(emptyList())
```

Comment:
The comment on this line is misleading. It is not what this code _does_. It does the same as `createDispatcher()` above, but without try/catch block around it. So if main dispatcher fails to intialize then on every invocation it does:
1. Invoke `createDispatcher`, which calls `mainFactory.createDispatcher(emptyList()) ` catch exception inside, replace it with null.
2. Invoke `tryCreateDispatcher` which again calls `mainFactory.createDispatcher(emptyList())` and now now throw its exception.